test/tools/d_do_test.d

```diff
@@ -479,28 +479,81 @@ bool collectExtraSources (in string input_dir, in string output_dir, in string[]
 {
-    import std.ascii : digits;
-    import std.utf : byCodeUnit;
+    import std.ascii : digits, fullHexDigits;
     for ( ; ; )
     {
-        auto pos = refoutput.indexOf("$n$");
-        if (pos < 0)
-            return refoutput == output;
-        if (output.length < pos)
+        auto npos = refoutput.indexOf("$n$");
+        auto hpos = refoutput.indexOf("$h$");
+        if (npos >= 0 && npos < cast(size_t)hpos)
+        {
+            if (matchUpToPosition(output, refoutput, npos, digits))
+                continue;
             return false;
-        if (refoutput[0..pos] != output[0..pos])
+        }
+        if (hpos >= 0)
+        {
+            if (matchUpToPosition(output, refoutput, hpos, fullHexDigits))
+                continue;
             return false;
-        refoutput = refoutput[pos + 3 ..$];
-        output = output[pos..$];
-        auto p = output.byCodeUnit.countUntil!(e => !digits.canFind(e));
-        output = output[p..$];
+        }
+        return refoutput == output;
     }
 }
 
+bool matchUpToPosition(ref string output, ref string refoutput, size_t pos, string match)
+{
+    import std.utf : byCodeUnit;
+    if (output.length < pos)
+        return false;
+    if (refoutput[0..pos] != output[0..pos])
+        return false;
+    if (!match.canFind(output[pos]))
+        return false;
+    refoutput = refoutput[pos + 3 .. $];
+    output = output[pos..$];
+    auto p = output.byCodeUnit.countUntil!(e => !match.canFind(e));
+    output = output[p < 0 ? $ : p .. $];
+    return true;
+}
+
+unittest
+{
+    assert( compareOutput(`plain text`, `plain text`));
+    assert(!compareOutput(`plain text`, `plain t3xt`));
+    assert( compareOutput(`number 1029384756`, `number $n$`));
+    assert(!compareOutput(`number 1029384756`, `$n$ number`));
+    assert( compareOutput(`hex 0x1f2e3d4c5b6a7089`, `hex 0x$h$`));
+    assert(!compareOutput(`hex 0x1f2e3d4c5b6a7089`, `0x$h$ hex`));
+
+    assert( compareOutput(`number 123 hex 0xabc`, `number $n$ hex 0x$h$`));
+    assert( compareOutput(`number 123 hex 0xabc`, `number $n$ hex $n$x$h$`));
+    assert(!compareOutput(`number 123 hex 0xabc`, `number $h$ hex $n$`));
+    assert( compareOutput(`hex 0xabc number 123`, `hex 0x$h$ number $n$`));
+    assert( compareOutput(`hex 0xabc number 123`, `hex $n$x$h$ number $n$`));
+    assert(!compareOutput(`hex 0xabc number 123`, `hex 0x$n$ number $h$`));
+
+    assert( compareOutput(`number 10UL`, `number $n$UL`));
+    assert(!compareOutput(`number UL`, `number $n$UL`));
+    assert(!compareOutput(`message`, `longer message`));
+    assert(!compareOutput(`message 123`, `longer message $n$`));
+    assert(!compareOutput(`unrecognized 123`, `unrecognized $N$`));
+    assert(!compareOutput(`unrecognized 0x123`, `unrecognized 0x$H$`));
+
+    assert( compareOutput(`123abc`, `$n$$h$`));
+    assert( compareOutput(`123n$`, `$n$n$`));
+    assert( compareOutput(`123n123`, `$n$n$n$`));
+    assert( compareOutput(`123n123n$`, `$n$n$n$n$`));
+}
```

Comment:
FWIW it's worth these unittest isn't run by any CI. We should really add a simple call to it (though that can happen in a different PR and I can open that one tomorrow).

Reply:
Good to know. :-)